### PR TITLE
ROX-11587: Prevent port 0 from being used as a server

### PR DESCRIPTION
**Pull request by Molter73**
### Description

Collector sometimes doesn't get the port number of some sockets, this may be due to it missing some syscall or the kernel simply not filling in the data. This last case most commonly happens with connectionless UDP messaging, where the kernel leaves the side that starts the communication empty.

In any case, we shouldn't be showing port 0 in the network graph, so this patch will make it so port 0 takes the least precedence, so it will only show up if both sides of the connection are 0.

### User-facing documentation

- [ ] CHANGELOG is updated **OR** update is not needed
- [x] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

### Testing and quality

- [x] the change is production ready: the change is GA or otherwise the functionality is gated by a feature flag
- [ ] CI results are inspected

#### Automated testing

- [x] added unit tests
- [ ] added e2e tests
- [ ] added regression tests
- [ ] added compatibility tests
- [ ] modified existing tests

#### How I validated my change

Added unit tests to ensure port 0 is not picked when using UDP.
Currently working on some changes for OCP Virt integration that make port 0 show up consistently, fixed with this change.

### CODE REVIEW

**Comment by JoukoVirtanen** [comment]
Please reenable the "Verify ports are greater than 0" test in `qa-tests-backend/src/test/groovy/NetworkFlowTest.groovy`.

File: sensor/common/networkflow/manager/manager_impl.go [hunk 1]
```diff
@@ -1094,49 +1094,64 @@ func getIPAndPort(address *sensor.NetworkAddress) net.NetworkPeerID {
...
-
-		// Special handling for UDP ports - role reported by collector may be unreliable, so look at which port is more
-		// likely to be ephemeral.
-		if remote.L4Proto == net.UDP {
-			incoming = netutil.IsEphemeralPort(remote.IPAndPort.Port) > netutil.IsEphemeralPort(local.Port)
+		var connType string
+		if c.incoming {
+			connType = "incoming"
+		} else {
+			connType = "outgoing"
```
Comment (Maddosaurus) [inline sensor/common/networkflow/manager/manager_impl.go:1144]: We can simplify this to
```suggestion
		connType := "outgoing"
		if c.incoming {
			connType = "incoming"
		}	
```

File: sensor/common/networkflow/manager/manager_impl_test.go [hunk 2]
```diff
@@ -1122,3 +1122,104 @@ func Test_connection_IsExternal(t *testing.T) {
...
+			},
+			RemoteAddress: &sensor.NetworkAddress{
+				Port: tt.remotePort,
+			},
+			Protocol: tt.protocol,
+			Role:     tt.role,
+		}
+		conn := getConnection(&networkConn)
+		assert.NotNil(t, conn)
+		assert.Equal(t, tt.expected, conn.incoming, "local: %d, remote: %d, protocol: %s, role: %s", tt.localPort, tt.remotePort, storage.L4Protocol_name[int32(tt.protocol)], sensor.ClientServerRole_name[int32(tt.role)])
```
Comment (Maddosaurus) [inline sensor/common/networkflow/manager/manager_impl_test.go:1223]: Maybe not for this test, but a new one: Wouldn't we also be interested in the correctness of the port, or is it only the directionality we're interested in?

File: sensor/common/networkflow/manager/manager_impl.go [hunk 3]
```diff
@@ -1094,49 +1094,64 @@ func getIPAndPort(address *sensor.NetworkAddress) net.NetworkPeerID {
 	return tuple
 }
 
+func getConnection(conn *sensor.NetworkConnection) *connection {
```
Comment (Maddosaurus) [inline sensor/common/networkflow/manager/manager_impl.go:1097]: I feel we can do better with the function name or the param types here:
It's called `getConnection`, takes a `NetworkConnection` and returns a `connection`.
For someone not intimately familiar with NWG or Collector, I am having trouble instantly grasping what this function does. 
We're determining directionality and extracting some information from a  `NetworkConnection`  (source & destination IP/Port , involved container), but that's something different that getting a connection, no?

File: sensor/common/networkflow/manager/manager_impl.go [hunk 4]
```diff
@@ -1094,49 +1094,64 @@ func getIPAndPort(address *sensor.NetworkAddress) net.NetworkPeerID {
...
+	// likely to be ephemeral. In case a port is set to 0, collector couldn't retrieve this value, we assume the
+	// connection works in the direction opposite of this port.
+	if remote.L4Proto == net.UDP {
+		incoming = netutil.IsEphemeralPort(remote.IPAndPort.Port) > netutil.IsEphemeralPort(local.Port)
+	}
+
+	return &connection{
+		local:       local,
+		remote:      remote,
+		containerID: conn.GetContainerId(),
```
Comment (Maddosaurus) [inline sensor/common/networkflow/manager/manager_impl.go:1124]: I suppose this container is either the receiving or sending end of a connection, right?

File: sensor/common/networkflow/manager/manager_impl.go [hunk 5]
```diff
@@ -1094,49 +1094,64 @@ func getIPAndPort(address *sensor.NetworkAddress) net.NetworkPeerID {
...
-		switch conn.Role {
-		case sensor.ClientServerRole_ROLE_SERVER:
-			flowMetrics.NetworkFlowsPerNodeByType.With(prometheus.Labels{"Hostname": hostname, "Type": "incoming", "Protocol": conn.Protocol.String()}).Inc()
-			incoming = true
-		case sensor.ClientServerRole_ROLE_CLIENT:
-			flowMetrics.NetworkFlowsPerNodeByType.With(prometheus.Labels{"Hostname": hostname, "Type": "outgoing", "Protocol": conn.Protocol.String()}).Inc()
-			incoming = false
-		default:
+		c := getConnection(conn)
+		if c == nil {
```
[the remark this answers is not in the document]
Comment (Molter73) [reply sensor/common/networkflow/manager/manager_impl.go:1136]: I think this might be more common than we might expect it to be, collector shouldn't be sending connections that are not marked as either server or client but I was seeing some unknowns when I was doing some testing. I'll check that it doesn't swamp the logs with garbage and implement one of the proposals if that is the case.

File: pkg/netutil/ephemeral_port.go [hunk 6]
```diff
@@ -17,6 +19,10 @@ const (
 // is the client by determining for which port there is the higher confidence that it is ephemeral.
 func IsEphemeralPort(port uint16) EphemeralPortConfidence {
 	switch {
+	// Port 0 means collector failed to get the port number, this happens
+	// in connectionless UDP messages sometimes.
+	case port == 0:
```
[the remark this answers is not in the document]
Comment (Molter73) [reply pkg/netutil/ephemeral_port.go:24]: This _should_ only happen for UDP, but I'm not 100% sure. In any case, this method is only used once and is in the `getConnection` method I created, specifically with UDP connections, it's not used at all with TCP. I'm happy moving the check outside of this function, the code just looked cleaner with this implementation.

File: sensor/common/networkflow/manager/manager_impl_test.go [hunk 7]
```diff
@@ -1122,3 +1122,104 @@ func Test_connection_IsExternal(t *testing.T) {
...
+			protocol:   storage.L4Protocol_L4_PROTOCOL_TCP,
+			role:       sensor.ClientServerRole_ROLE_CLIENT,
+			expected:   false,
+		}, {
+			localPort:  53,
+			remotePort: 50000,
+			protocol:   storage.L4Protocol_L4_PROTOCOL_TCP,
+			role:       sensor.ClientServerRole_ROLE_SERVER,
+			expected:   true,
+		},
```
[the remark this answers is not in the document]
Comment (Molter73) [reply sensor/common/networkflow/manager/manager_impl_test.go:1206]: It really shouldn't, if both the local and remote ports are marked as 0 we've dropped so much data that the connection is basically meaningless, it's more likely Falco would drop the event altogether. I'll add a test case and leave a comment explaining this situation if that's OK.

File: sensor/common/networkflow/manager/manager_impl_test.go (hunk 2 above)
Comment (Molter73) [reply sensor/common/networkflow/manager/manager_impl_test.go:1223]: Well, what would be the definition of correctness? 

Strictly speaking, a port is just a number, any port can be listened on, there's no right or wrong, there's only some reserved ports on the lower ranges for specific applications and some guidelines OSes use for assigning random ports when an application doesn't specify it. Other than that, anything is fair game, port 8080 is usually used for an application to listen on, but nothing prevents you from using that same port for sending data if it's unused.

File: sensor/common/networkflow/manager/manager_impl.go (hunk 3 above)
Comment (Molter73) [reply sensor/common/networkflow/manager/manager_impl.go:1097]: Yeah, I know, it's terrible. I'll figure out a better name for it.

File: sensor/common/networkflow/manager/manager_impl.go (hunk 4 above)
Comment (Molter73) [reply sensor/common/networkflow/manager/manager_impl.go:1124]: I believe it's the container the `local` side is on, but the code in collector is a bit of a tangled mess with calls to Falco, I might be wrong.

File: pkg/netutil/ephemeral_port.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Maddosaurus) [reply pkg/netutil/ephemeral_port.go:24]: I am not too well versed in this area of our code, so I'll trust your intuition. I agree, it does look clean, and if you don't think this is going to result in masking errors, I'd be fine with leaving it as is

File: sensor/common/networkflow/manager/manager_impl_test.go (hunk 2 above)
Comment (Maddosaurus) [reply sensor/common/networkflow/manager/manager_impl_test.go:1223]: I see, my question was too broad, let me try again :sweat_smile: 
We're checking for the correct value of `incoming`, but we don't look at the local or remote ports. 
Is that not important for this function, or any uses of the returned struct?  If not, my comment is a non-issue

Comment (Molter73) [reply sensor/common/networkflow/manager/manager_impl_test.go:1223]: Ah, I see. Well the translation of the port is [simple enough](https://github.com/stackrox/stackrox/blob/73c7beec618fd5ebdaf6bbaa3e04e3833448380d/sensor/common/networkflow/manager/manager_impl.go#L1092), so I don't see much value in testing that. The ip addresses are a bit more involved, but those already have tests [here](https://github.com/stackrox/stackrox/blob/master/pkg/net/addr_test.go), so it feels like I'd be sort of duplicating those. WDYT? Is there a specific test case you have in mind?

File: sensor/common/networkflow/manager/manager_impl.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Molter73) [reply sensor/common/networkflow/manager/manager_impl.go]: Ended up adding an extra error return value and logging it in the caller with `Warn` severity, didn't see it triggering after some quick testing, so we are probably fine. Happy to change the severity if that one is not the correct one.

File: sensor/common/networkflow/manager/manager_impl.go (hunk 3 above)
Comment (Molter73) [reply sensor/common/networkflow/manager/manager_impl.go]: Was torn between `cookConnection` and `processConnection`, went with the boring option for "serious points", which is totally not a thing I just made up.

**Review by vikin91 — COMMENTED** [review]
I think the change looks fine. I understand that the ephemeral nature of the port has only influence on the direction of the connection - would that help not showing the `0` on the graph?

**Comment by Molter73** [comment]
> I think the change looks fine. I understand that the ephemeral nature of the port has only influence on the direction of the connection - would that help not showing the `0` on the graph?

I'm currently working on integration with OCP Virt, when using UDP with some VMs I was getting port 0 everywhere, this change got rid of it altogether. Granted, it's not the exact error the user is having, but I believe it should fix it for them as well.

The only corner case this could cause is that we incorrectly show a client port as a server if we fail to get the port number for the server, but either way would be wrong and this is (only slightly) more correct, since we don't show a completely wrong port.

**Review by vikin91 — APPROVED** [review]
Thanks! Green from me, but you still need to look into that style-check that failed, I do not understand why it complains here about some scanner code - maybe a rebase is necessary?

**Comment by Molter73** [comment]
> Thanks! Green from me, but you still need to look into that style-check that failed, I do not understand why it complains here about some scanner code - maybe a rebase is necessary?

Yeah, it was complaining about a test that I hadn't change, a rebase fixed it.

File: sensor/common/networkflow/manager/manager_impl_test.go (hunk 2 above)
Comment (Maddosaurus) [reply sensor/common/networkflow/manager/manager_impl_test.go:1239]: No, that's great as-is, thanks!

File: sensor/common/networkflow/manager/manager_impl.go (hunk 3 above)
Comment (Maddosaurus) [reply sensor/common/networkflow/manager/manager_impl.go]: Ahh, that's what being a professional:tm: means, I see :laughing: 
Thanks for improving!

**Review by Maddosaurus — APPROVED** [review]
:rocket: 

Final thought: Maybe check the failing gke-qa-e2e test ([here](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/stackrox_stackrox/12332/pull-ci-stackrox-stackrox-master-gke-qa-e2e-tests/1821914027060629504)), as it seems to contain an unexpected collector DaemonSet restart :thinking:

**Review by robbycochran — APPROVED** [review]
